File: library/src/main/java/com/hsicen/library/java/FanLayoutManagerSettings.java
package com.hsicen.library.java;

import android.content.Context;

/**
 * @author alex yarovoi
 * @version 1.0
 */
public class FanLayoutManagerSettings {

    private static final float DEFAULT_VIEW_WIDTH_DP = 120F;
    private static final float DEFAULT_VIEW_HEIGHT_DP = 160F;

    private float mViewWidthDp;
    private float mViewHeightDp;
    private int mViewWidthPx;
    private int mViewHeightPx;
    private boolean mIsFanRadiusEnable;
    private int mItemMargin;
    private boolean mClickScale;

    public int getItemMargin() {
        return mItemMargin;
    }

    public boolean isClickScale() {
        return mClickScale;
    }

    public void setClickScale(boolean mClickScale) {
        this.mClickScale = mClickScale;
    }

    public void setItemMargin(int mItemMargin) {
        this.mItemMargin = mItemMargin;
    }

    public int getTopMargin() {
        return mTopMargin;
    }

    public void setTopMargin(int mTopMargin) {
        this.mTopMargin = mTopMargin;
    }

    public int getBottomMargin() {
        return mBottomMargin;
    }

    public void setBottomMargin(int mBottomMargin) {
        this.mBottomMargin = mBottomMargin;
    }

    private int mTopMargin;
    private int mBottomMargin;

    private FanLayoutManagerSettings(Builder builder) {
        mViewWidthDp = builder.mViewWidthDp;
        mViewHeightDp = builder.mViewHeightDp;
        mIsFanRadiusEnable = builder.mIsFanRadiusEnable;
        mViewWidthPx = builder.mViewWidthPx;
        mViewHeightPx = builder.mViewHeightPx;
        mItemMargin = builder.mItemMargin;
        mTopMargin = builder.mTopMargin;
        mBottomMargin = builder.mBottomMargin;
        mClickScale = builder.mClickScale;
    }

    public static Builder newBuilder(Context context) {
        return new Builder(context);
    }

    float getViewWidthDp() {
        return mViewWidthDp;
    }

    float getViewHeightDp() {
        return mViewHeightDp;
    }

    boolean isFanRadiusEnable() {
        return mIsFanRadiusEnable;
    }

    int getViewWidthPx() {
        return mViewWidthPx;
    }

    int getViewHeightPx() {
        return mViewHeightPx;
    }

    /**
     * {@code FanLayoutManagerSettings} builder static inner class.
     */
    public static final class Builder {
        private static final float BOUNCE_MAX = 10F;
        private Context mContext;
        private float mViewWidthDp;
        private float mViewHeightDp;
        private boolean mIsFanRadiusEnable;
        private int mViewWidthPx;
        private int mViewHeightPx;
        private int mItemMargin;
        private int mTopMargin;
        private int mBottomMargin;
        private boolean mClickScale;

        private Builder(Context context) {
            mContext = context;
        }

        /**
         * Sets the {@code mViewWidthDp} and returns a reference to this Builder so that the methods can be chained together.
         *
         * @param viewWidthDp the {@code mViewWidthDp} to set
         * @return a reference to this Builder
         */
        public Builder withViewWidthDp(float viewWidthDp) {
            mViewWidthDp = viewWidthDp;
            mViewWidthPx = Math.round(mContext.getResources().getDisplayMetrics().density * viewWidthDp);
            mViewWidthPx = Math.min(mContext.getResources().getDisplayMetrics().widthPixels, mViewWidthPx);
            return this;
        }

        /**
         * Sets the {@code mViewHeightDp} and returns a reference to this Builder so that the methods can be chained together.
         *
         * @param viewHeightDp the {@code mViewHeightDp} to set
         * @return a reference to this Builder
         */
        public Builder withViewHeightDp(float viewHeightDp) {
            mViewHeightDp = viewHeightDp;
            mViewHeightPx = Math.round(mContext.getResources().getDisplayMetrics().density * viewHeightDp);
            mViewHeightPx = Math.min(mContext.getResources().getDisplayMetrics().heightPixels, mViewHeightPx);
            return this;
        }

        /**
         * Sets the {@code fanRadius} and returns a reference to this Builder so that the methods can be chained together.
         *
         * @param isFanRadiusEnable the {@code mIsFanRadiusEnable} to set
         * @return a reference to this Builder
         */
        public Builder withFanRadius(boolean isFanRadiusEnable) {
            mIsFanRadiusEnable = isFanRadiusEnable;
            return this;
        }


        public Builder withItemMargin(int itemMargin) {
            if (itemMargin <= 0F) {
                return this;
            }

            mItemMargin = itemMargin;
            return this;
        }

        public Builder withTopMargin(int topMargin) {
            if (topMargin <= 0F) {
                return this;
            }

            mTopMargin = topMargin;
            return this;
        }

        public Builder withBottomMargin(int bottomMargin) {
            if (bottomMargin <= 0F) {
                return this;
            }

            mBottomMargin = bottomMargin;
            return this;
        }

        public Builder withClickScale(boolean clickScale) {
            mClickScale = clickScale;
            return this;
        }

        /**
         * Returns a {@code FanLayoutManagerSettings} built from the parameters previously set.
         *
         * @return a {@code FanLayoutManagerSettings} built with parameters of this {@code FanLayoutManagerSettings.Builder}
         */
        public FanLayoutManagerSettings build() {
            if (Float.compare(mViewWidthDp, 0F) == 0) {
                withViewWidthDp(DEFAULT_VIEW_WIDTH_DP);
            }
            if (Float.compare(mViewHeightDp, 0F) == 0) {
                withViewHeightDp(DEFAULT_VIEW_HEIGHT_DP);
            }
            return new FanLayoutManagerSettings(this);
        }
    }
}

File: library/src/main/java/com/hsicen/library/FanLayoutManagerSettings.kt
package com.hsicen.library

import android.content.Context
import kotlin.math.min
import kotlin.math.roundToInt

/**
 * 作者：hsicen  2020/10/2 16:21
 * 邮箱：codinghuang@163.com
 * 功能：
 * 描述：FanLayoutManager配置参数
 */
class FanLayoutManagerSettings private constructor(builder: Builder) {
  val viewWidthDp: Float
  val viewHeightDp: Float
  val viewWidthPx: Int
  val viewHeightPx: Int
  val isFanRadiusEnable: Boolean
  val angleItemBounce: Float

  class Builder constructor(private val mContext: Context) {
    var mViewWidthDp = 0f
    var mViewHeightDp = 0f
    var mIsFanRadiusEnable = false
    var mAngleItemBounce = 0f
    var mViewWidthPx = 0
    var mViewHeightPx = 0

    /**
     * 设置宽度
     * @param viewWidthDp dp值
     */
    fun withViewWidthDp(viewWidthDp: Float): Builder {
      mViewWidthDp = viewWidthDp
      mViewWidthPx = (mContext.resources.displayMetrics.density * viewWidthDp).roundToInt()
      mViewWidthPx = min(mContext.resources.displayMetrics.widthPixels, mViewWidthPx)
      return this
    }


    /**
     * 设置高度
     * @param viewHeightDp dp值
     */
    fun withViewHeightDp(viewHeightDp: Float): Builder {
      mViewHeightDp = viewHeightDp
      mViewHeightPx = (mContext.resources.displayMetrics.density * viewHeightDp).roundToInt()
      mViewHeightPx = min(mContext.resources.displayMetrics.heightPixels, mViewHeightPx)
      return this
    }

    /**
     * 设置是否启用扇形布局
     * @param isFanRadiusEnable 设置是否启用扇形布局
     */
    fun withFanRadius(isFanRadiusEnable: Boolean): Builder {
      mIsFanRadiusEnable = isFanRadiusEnable
      return this
    }

    /**
     * 设置扇形布局角度
     * @param angleItemBounce 扇形角度
     */
    fun withAngleItemBounce(angleItemBounce: Float): Builder {
      if (angleItemBounce <= 0f) {
        return this
      }
      mAngleItemBounce = min(BOUNCE_MAX, angleItemBounce)
      return this
    }


    fun build(): FanLayoutManagerSettings {
      if (mViewWidthDp.compareTo(0f) == 0) {
        withViewWidthDp(DEFAULT_VIEW_WIDTH_DP)
      }
      if (mViewHeightDp.compareTo(0f) == 0) {
        withViewHeightDp(DEFAULT_VIEW_HEIGHT_DP)
      }
      return FanLayoutManagerSettings(this)
    }

    companion object {
      private const val BOUNCE_MAX = 10f
    }
  }

  companion object {
    private const val DEFAULT_VIEW_WIDTH_DP = 120f
    private const val DEFAULT_VIEW_HEIGHT_DP = 160f

    fun newBuilder(context: Context): Builder {
      return Builder(context)
    }
  }

  init {
    viewWidthDp = builder.mViewWidthDp
    viewHeightDp = builder.mViewHeightDp
    isFanRadiusEnable = builder.mIsFanRadiusEnable
    angleItemBounce = builder.mAngleItemBounce
    viewWidthPx = builder.mViewWidthPx
    viewHeightPx = builder.mViewHeightPx
  }
}
